src/ActionList/Visuals.tsx

```diff
@@ -75,3 +77,56 @@ export const TrailingVisual: React.FC<React.PropsWithChildren<VisualProps>> = ({
     </Box>
   )
 }
+
+// VisualOrIndicator handles the positioning of indicators and determines whether to show a visual or an indicator.
+//
+// If we're showing an *inactive* or *loading* indicator and a leading visual has NOT been passed,
+// replace the trailing visual with the inactive indicator.
+//
+// This preserves the left alignment of item text.
+export const VisualOrIndicator: React.FC<
+  React.PropsWithChildren<{
+    inactiveText?: TooltipProps['text']
+    itemHasLeadingVisual: boolean
+    labelId?: string
+    loading?: boolean
+    position: 'leading' | 'trailing'
+  }>
+> = ({children, labelId, loading, inactiveText, itemHasLeadingVisual, position}) => {
+  const VisualComponent = position === 'leading' ? LeadingVisual : TrailingVisual
+
+  if (!loading && !inactiveText) return children
+
+  if (
+    (itemHasLeadingVisual && position === 'trailing') || // has a leading visual, and it's in the trailing position, or
+    (!itemHasLeadingVisual && position === 'leading') // it doesn't have a leading visual, and it's in the leading position
+  ) {
+    // => so we don't render the indicator here
+    return children
+  }
+
+  return inactiveText ? (
+    <Tooltip text={inactiveText}>
+      <Box
+        as="button"
```

Comment:
Just curious, are we using the `Tooltip` + `Button` combo here as a last option? I know this is an anti-pattern, so I just wanted to confirm!

Also, I see we're not using this pattern for when the `Spinner` is active. Is the difference here being that the `Spinner` state is temporary rather than permanent?

Reply:
> Just curious, are we using the Tooltip + Button combo here as a last option? I know this is an anti-pattern, so I just wanted to confirm!

Yes, this is the direction I got from a11y office hours: https://github.com/github/accessibility/issues/4779

---

> Also, I see we're not using this pattern for when the Spinner is active. Is the difference here being that the Spinner state is temporary rather than permanent?

Correct.